core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractGroupNode.java

```diff
@@ -63,23 +75,149 @@ public void insert(InTuple_ tuple) {
                     + ") was already added in the tupleStore.");
         }
         GroupKey_ groupKey = createGroupKey(tuple);
-        Group<OutTuple_, GroupKey_, ResultContainer_> group = groupMap.computeIfAbsent(groupKey,
-                k -> new Group<>(k, runAccumulate ? supplier.get() : null));
+        Group<OutTuple_, GroupKey_, ResultContainer_> group = groupMap.get(groupKey);
+        if (group == null) {
+            ResultContainer_ resultContainer = hasCollector ? supplier.get() : null;
+            OutTuple_ outTuple = createOutTuple(groupKey);
+            outTuple.setState(BavetTupleState.CREATING);
+            group = new Group<>(groupKey, resultContainer, outTuple);
+            groupMap.put(groupKey, group);
+            // Don't add it if (state == CREATING), but (group != null), which is a 2th insert of the same groupKey.
+            dirtyGroupQueue.add(group);
+        }
         group.parentCount++;
-
-        Runnable undoAccumulator = runAccumulate ? accumulate(group.resultContainer, tuple) : null;
+        Runnable undoAccumulator = hasCollector ? accumulate(group.resultContainer, tuple) : null;
         GroupPart<Group<OutTuple_, GroupKey_, ResultContainer_>> groupPart = new GroupPart<>(group, undoAccumulator);
         tupleStore[groupStoreIndex] = groupPart;
-        if (!group.dirty) {
-            group.dirty = true;
-            dirtyGroupQueue.add(group);
+
+        switch (group.outTuple.getState()) {
+            case CREATING:
+                break;
+            case UPDATING:
+                break;
+            case OK:
+                group.outTuple.setState(BavetTupleState.UPDATING);
+                dirtyGroupQueue.add(group);
+                break;
+            case DYING:
+                group.outTuple.setState(BavetTupleState.UPDATING);
+                break;
+            case ABORTING:
+                group.outTuple.setState(BavetTupleState.CREATING);
+                break;
+            case DEAD:
+            default:
+                throw new IllegalStateException("Impossible state: The group (" + group + ") in node (" +
+                        this + ") is in an unexpected state (" + group.outTuple.getState() + ").");
         }
     }
 
     public void update(InTuple_ tuple) {
```

Comment:
This method is very long, and its control flow confusing. Ifs, switches, breaks, ...
Maybe split it into some submethods for functional blocks, so that it becomes easier to reason about what happens when?